File: src/main/session.ts
/**
 * 登录态存储。
 *
 * 只保存站点的会话 cookie（PHPSESSID），**不保存账号密码** ——
 * 用户在浏览器里自己登录，再把 cookie 交给应用。
 *
 * 落盘用 Electron 的 safeStorage（Windows 下走 DPAPI，绑定当前用户账户），
 * 明文不写文件。若当前环境拿不到加密能力，就退化成「只存内存」，
 * 宁可每次重贴，也不把凭证明文写到磁盘上。
 */
import { existsSync } from 'node:fs'
import { readFile, rm, writeFile } from 'node:fs/promises'
import { join } from 'node:path'
import { app, safeStorage } from 'electron'

export interface SessionStatus {
  loggedIn: boolean
  /** 凭据指纹（只用于界面显示，不可反推），未登录为 null */
  fingerprint: string | null
  savedAt: string | null
  /** 能否加密落盘 */
  encrypted: boolean
  /** 站点校验结果：null 表示还没验过 */
  verified: boolean | null
  verifyMessage: string | null
}

export class SessionStore {
  private readonly file: string
  private cookie: string | null = null
  private savedAt: string | null = null
  private verified: boolean | null = null
  private verifyMessage: string | null = null
  /** 会话失效只提醒一次，避免刷屏 */
  private expiryNotified = false

  constructor() {
    this.file = join(app.getPath('userData'), 'session.bin')
  }

  private get canEncrypt(): boolean {
    try {
      return safeStorage.isEncryptionAvailable()
    } catch {
      return false
    }
  }

  async load(): Promise<void> {
    if (!existsSync(this.file)) return
    try {
      const raw = await readFile(this.file)
      if (!this.canEncrypt) return
      const text = safeStorage.decryptString(raw)
      const parsed = JSON.parse(text) as { cookie?: string; savedAt?: string }
      if (parsed.cookie) {
        this.cookie = parsed.cookie
        this.savedAt = parsed.savedAt ?? null
      }
    } catch {
      // 解不开（换了机器 / 换了用户）就当没有
      await this.clear()
    }
  }

  /** 当前要注入请求的 Cookie 头，未登录返回 null */
  get cookieHeader(): string | null {
    if (!this.cookie) return null
    return this.cookie.includes('=') ? this.cookie : `PHPSESSID=${this.cookie}`
  }

  isLoggedIn(): boolean {
    return this.cookieHeader != null
  }

  /** 提醒过一次就不再提醒 */
  shouldNotifyExpiry(): boolean {
    return !this.expiryNotified
  }

  markExpired(message: string): void {
    this.verified = false
    this.verifyMessage = message
    this.expiryNotified = true
  }

  markVerified(): void {
    this.verified = true
    this.verifyMessage = null
    this.expiryNotified = false
  }

  async save(input: string): Promise<void> {
    const trimmed = input.trim()
    if (!trimmed) throw new Error('cookie 不能为空')
    this.cookie = trimmed
    this.savedAt = new Date().toISOString()
    this.verified = null
    this.verifyMessage = null
    this.expiryNotified = false
    if (!this.canEncrypt) return
    const payload = JSON.stringify({ cookie: this.cookie, savedAt: this.savedAt })
    await writeFile(this.file, safeStorage.encryptString(payload))
  }

  /**
   * 跟着站点换 cookie。
   *
   * 浏览器之所以能一直保持登录，是因为它每次都按 Set-Cookie 更新本地值；
   * 应用只存一份静态 cookie 的话，站点一换就是废票。这里做同样的事，
   * 并且立刻重新加密落盘，免得重启后又用回旧值。
   */
  async adopt(setCookie: string): Promise<boolean> {
    // 只有本来就登录着才跟着换。
    // 站点给每个访客都会发一个 PHPSESSID，没有这道闸门的话，
    // 匿名抓一把就会把应用「变成」已登录，泳装分享分类也会莫名其妙冒出来。
    if (!this.cookie) return false
    // 明确作废的 Set-Cookie 不能当成新会话
    if (/(max-age=0|expires=thu,\s*01-jan-1970)/i.test(setCookie)) return false
    const match = /PHPSESSID=([^;,\s]*)/i.exec(setCookie)
    const value = (match?.[1] ?? '').trim()
    if (!value || value.toLowerCase() === 'deleted') return false
    const current = this.cookie?.includes('=') ? this.cookie.split('=').slice(1).join('=') : this.cookie
    if (current === value) return false
    this.cookie = `PHPSESSID=${value}`
    this.savedAt = new Date().toISOString()
    // 站点刚刚回过话，说明这份会话是活的
    this.verified = true
    this.verifyMessage = null
    this.expiryNotified = false
    if (!this.canEncrypt) return true
    const payload = JSON.stringify({ cookie: this.cookie, savedAt: this.savedAt })
    await writeFile(this.file, safeStorage.encryptString(payload)).catch(() => {})
    return true
  }

  async clear(): Promise<void> {
    this.cookie = null
    this.savedAt = null
    this.verified = null
    this.verifyMessage = null
    this.expiryNotified = false
    await rm(this.file, { force: true }).catch(() => {})
  }

  status(): SessionStatus {
    return {
      loggedIn: this.isLoggedIn(),
      fingerprint: this.cookie ? fingerprint(this.cookie) : null,
      savedAt: this.savedAt,
      encrypted: this.canEncrypt,
      verified: this.verified,
      verifyMessage: this.verifyMessage
    }
  }
}

/** 只暴露头尾几位，用来确认「是不是换了一份」，不足以还原凭证 */
function fingerprint(cookie: string): string {
  const value = cookie.includes('=') ? cookie.split('=').slice(1).join('=') : cookie
  if (value.length <= 8) return value
  return `${value.slice(0, 4)}…${value.slice(-4)}`
}

File: src-tauri/src/session.rs
/**
 * 登录态：只保存站点会话 cookie，**不保存账号密码**。
 *
 * 与 Electron 版的 src/main/session.ts 语义一致，差别是落盘方式：
 * 旧版写 userData/session.bin（DPAPI 密文）；新版交给系统凭据库
 * （Windows 凭据管理器，同样是 DPAPI 保护、绑定当前用户账户）。
 */
use crate::crawler::http::{HttpClient, HttpOptions};
use crate::crawler::site::SITE_ORIGIN;
use serde_json::{json, Value as Json};

const SERVICE: &str = "gugu-gallery";
const ACCOUNT: &str = "session-cookie";

/// 自检用：环境变量 GUGU_SESSION_EPHEMERAL=1 时不读写系统凭据库，
/// 保证断言与用户本机的登录状态无关。
fn ephemeral() -> bool {
    std::env::var("GUGU_SESSION_EPHEMERAL").map(|v| v == "1").unwrap_or(false)
}

pub struct SessionStore {
    cookie: Option<String>,
    saved_at: Option<String>,
    verified: Option<bool>,
    verify_message: Option<String>,
    expiry_notified: bool,
}

impl Default for SessionStore {
    fn default() -> Self {
        Self::new()
    }
}

impl SessionStore {
    pub fn new() -> Self {
        Self {
            cookie: None,
            saved_at: None,
            verified: None,
            verify_message: None,
            expiry_notified: false,
        }
    }

    fn entry() -> Option<keyring::Entry> {
        keyring::Entry::new(SERVICE, ACCOUNT).ok()
    }

    /// 启动时读回登录态
    pub fn load(&mut self) {
        if ephemeral() {
            return;
        }
        let Some(entry) = Self::entry() else { return };
        let Ok(raw) = entry.get_password() else { return };
        if let Ok(parsed) = serde_json::from_str::<Json>(&raw) {
            let cookie = parsed.get("cookie").and_then(|v| v.as_str()).map(|s| s.to_string());
            if let Some(cookie) = cookie.filter(|c| !c.is_empty()) {
                self.saved_at = parsed.get("savedAt").and_then(|v| v.as_str()).map(|s| s.to_string());
                self.cookie = Some(cookie);
            }
        }
    }

    pub fn is_logged_in(&self) -> bool {
        self.cookie_header().is_some()
    }

    pub fn cookie_header(&self) -> Option<String> {
        let cookie = self.cookie.as_deref()?;
        if cookie.is_empty() {
            return None;
        }
        if cookie.contains('=') {
            Some(cookie.to_string())
        } else {
            Some(format!("PHPSESSID={cookie}"))
        }
    }

    pub fn save(&mut self, input: &str) -> Result<(), String> {
        let trimmed = input.trim();
        if trimmed.is_empty() {
            return Err("cookie 不能为空".into());
        }
        self.cookie = Some(trimmed.to_string());
        self.saved_at = Some(chrono::Local::now().to_rfc3339());
        self.verified = None;
        self.verify_message = None;
        self.expiry_notified = false;
        if !ephemeral() {
          if let Some(entry) = Self::entry() {
            let payload = json!({ "cookie": trimmed, "savedAt": self.saved_at }).to_string();
            // 凭据库不可用时退化为「仅本次有效」，绝不落明文
            let _ = entry.set_password(&payload);
          }
        }
        Ok(())
    }

    pub fn clear(&mut self) {
        self.cookie = None;
        self.saved_at = None;
        self.verified = None;
        self.verify_message = None;
        self.expiry_notified = false;
        if !ephemeral() {
          if let Some(entry) = Self::entry() {
            let _ = entry.delete_credential();
          }
        }
    }

    pub fn status(&self) -> Json {
        json!({
            "loggedIn": self.is_logged_in(),
            "fingerprint": self.cookie.as_deref().map(fingerprint),
            "savedAt": self.saved_at,
            "encrypted": true,
            "verified": self.verified,
            "verifyMessage": self.verify_message
        })
    }

    pub fn mark_verified(&mut self) {
        self.verified = Some(true);
        self.verify_message = None;
        self.expiry_notified = false;
    }

    pub fn mark_expired(&mut self, message: &str) {
        self.verified = Some(false);
        self.verify_message = Some(message.to_string());
        self.expiry_notified = true;
    }

    pub fn should_notify_expiry(&self) -> bool {
        !self.expiry_notified
    }

    /// 把一次校验结果落到会话状态上（不碰 IO，可安全地在锁内调用）
    pub fn apply_verify(&mut self, state: &str, message: &str) {
        if state == "in" {
            self.mark_verified();
        } else if state == "out" {
            self.mark_expired(message);
        } else {
            self.verified = Some(false);
            self.verify_message = Some(message.to_string());
        }
    }

    /// 校验前先把要用的 cookie 取出来（避免把锁带进 async）
    pub fn cookie_for_verify(&self) -> Option<String> {
        self.cookie_header()
    }
}

/// 去站点核对一次：不持锁、不依赖 SessionStore，方便在 async 命令里调用
pub async fn verify_cookie(cookie: Option<String>, proxy: Option<String>) -> (Json, String, String) {
    let Some(cookie) = cookie else {
        return (
            json!({ "ok": false, "state": "none", "message": "还没有设置登录凭据" }),
            "none".into(),
            "还没有设置登录凭据".into(),
        );
    };
    let client = HttpClient::new(HttpOptions {
        delay_ms: 0,
        retries: 1,
        timeout_ms: 20_000,
        concurrency: 1,
        cookie: Some(cookie),
        proxy,
    });
    match client.get_html(&format!("{SITE_ORIGIN}/")).await {
        Ok(html) => match login_state(&html) {
            "in" => (json!({ "ok": true, "state": "in", "message": "登录态有效" }), "in".into(), "登录态有效".into()),
            "out" => {
                let message = "cookie 已过期，请重新登录";
                (json!({ "ok": false, "state": "out", "message": message }), "out".into(), message.into())
            }
            _ => {
                let message = "没法确认登录态：首页里既没有登录入口也没有退出入口";
                (json!({ "ok": false, "state": "unknown", "message": message }), "unknown".into(), message.into())
            }
        },
        Err(err) => {
            let message = format!("校验失败：{err}");
            (json!({ "ok": false, "state": "error", "message": message }), "error".into(), message)
        }
    }
}

/// 站点判定登录态：未登录时页头固定是「登录 / 注册」，登录后换成用户入口并多出「泳装类分享」
pub fn login_state(html: &str) -> &'static str {
    if html.contains("泳装类分享") {
        return "in";
    }
    if html.contains("退出") || html.contains("注销") {
        return "in";
    }
    if html.contains("登录 / 注册") || html.contains("/login.html") {
        return "out";
    }
    "unknown"
}

/// 只暴露头尾几位，够确认「换没换」但不足以还原凭证
fn fingerprint(cookie: &str) -> String {
    let value = if let Some(idx) = cookie.find('=') {
        &cookie[idx + 1..]
    } else {
        cookie
    };
    let chars: Vec<char> = value.chars().collect();
    if chars.len() <= 8 {
        return value.to_string();
    }
    let head: String = chars.iter().take(4).collect();
    let tail: String = chars.iter().rev().take(4).collect::<Vec<_>>().into_iter().rev().collect();
    format!("{head}…{tail}")
}
